helper/salt/salt.go

```diff
@@ -57,46 +73,91 @@ func NewSalt(view logical.Storage, config *Config) (*Salt, error) {
 		config: config,
 	}
 
-	// Look for the salt
-	raw, err := view.Get(config.Location)
-	if err != nil {
-		return nil, fmt.Errorf("failed to read salt: %v", err)
-	}
-
-	// Restore the salt if it exists
-	if raw != nil {
-		s.salt = string(raw.Value)
+	var raw *logical.StorageEntry
+	var err error
+	if config.StaticSalt != "" {
+		s.salt = config.StaticSalt
+	} else {
+		if view != nil {
+			// Look for the salt
+			raw, err = view.Get(config.Location)
+			if err != nil {
+				return nil, fmt.Errorf("failed to read salt: %v", err)
+			}
+
+			// Restore the salt if it exists
+			if raw != nil {
+				s.salt = string(raw.Value)
+			}
+		}
 	}
 
 	// Generate a new salt if necessary
 	if s.salt == "" {
 		s.salt = uuid.GenerateUUID()
 		s.generated = true
-		raw = &logical.StorageEntry{
-			Key:   config.Location,
-			Value: []byte(s.salt),
+		if view != nil {
+			raw = &logical.StorageEntry{
+				Key:   config.Location,
+				Value: []byte(s.salt),
+			}
+			if err := view.Put(raw); err != nil {
+				return nil, fmt.Errorf("failed to persist salt: %v", err)
+			}
+		}
+	}
+
+	if config.HMAC != nil {
+		if len(config.HMACType) == 0 {
+			return nil, fmt.Errorf("HMACType must be defined")
 		}
-		if err := view.Put(raw); err != nil {
-			return nil, fmt.Errorf("failed to persist salt: %v", err)
+		s.hmac = hmac.New(config.HMAC, []byte(s.salt))
```

Comment:
The HMAC construction cannot be shared I'm fairly sure. I think its done in one-shot